### PR TITLE
Disallow ALTER TABLE ADD CONSTRAINT USING INDEX on a partition root

### PR DESCRIPTION
**Pull request by bhuvnesh2703**
ADD CONSTRAINT USING INDEX is a new feature which didn't existing for 4 or 5 branches.
.This is taken from upstream commit eb7ed3f. With partition tables, the changes are not propagated to the leaf partitions if the operation is done on the root table, thus disallowed currently.

Signed-off-by: Soumyadeep Chakraborty <sochakraborty@pivotal.io>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/test/regress/sql/create_index.sql [hunk 1]
```diff
@@ -805,6 +805,12 @@ DROP INDEX cwi_replaced_pkey;	-- Should fail; a constraint depends on it
 
 DROP TABLE cwi_test;
 
+-- ADD CONSTRAINT USING INDEX is forbidden on partitioned tables
+CREATE TABLE cwi_test(a int) PARTITION BY range(a) (START(1) END(2) EVERY(1));
```
Comment (asimrp) [inline src/test/regress/sql/create_index.sql]: Nit-pick: can we please have distribution key separate from partition key?  Even if this is a test case, it should serve as an illustration of real life scenario as much as possible.

File: src/backend/commands/tablecmds.c [hunk 2]
```diff
@@ -8852,6 +8852,16 @@ ATExecAddIndexConstraint(AlteredTableInfo *tab, Relation rel,
 	Assert(OidIsValid(index_oid));
 	Assert(stmt->isconstraint);
 
+	/*
+	 * Doing this on partitioned tables is not a simple feature to implement,
+	 * so let's punt for now.
+	 */
+	Oid rel_id = RelationGetRelid(rel);
+	if (Gp_role != GP_ROLE_EXECUTE && rel_is_partitioned(rel_id))
```
Comment (asimrp) [inline src/backend/commands/tablecmds.c]: Looking at this condition, I went through 5X code to see how is this disallowed there.  Turned out, USING INDEX part of the alter table commands is a new feature in 6X and beyond that was inherited from PostgreSQL.  Can we please make it clear in the commit message, that this was not disallowed but not supported in older Greenplum versions?  That also means we don't have to worry about upgrade, yay!

Speaking of commit message, can we please use "Co-authored-by:" for authorship?  We should remove any remnants of variants such as "Signed-off-by" from local workspace once and for all.

**Comment by bhuvnesh2703** [comment]
> I hate to ask but should we be worried about exchange partition? What if the table being incorporated in the partition hierarchy contains a constraint defined using an index?

Thanks for the question, we tried it. It errors out stating that the constraint existing is invalid. See below.
```
create table t1 (a int, b int, c int) partition by range(b) (start(1) end(5) every(1));
CREATE TABLE
create table t1_exchange_part (a int, b int, c int) ;
CREATE TABLE
create unique index t2_btree_idx on t1_exchange_part(a);
CREATE INDEX
alter table t1_exchange_part  add constraint t2_asdfas unique using index t2_btree_idx;
ALTER TABLE t1  EXCHANGE PARTITION FOR (RANK(2)) WITH TABLE t1_exchange_part;
ERROR:  invalid constraint(s) found on "t1_exchange_part": "t2_asdfas"
HINT:  Drop the invalid constraints and retry.
```

**Comment by bhuvnesh2703** [comment]
@asimrp Applied the other feedback (updating comment, Co-authored-by and test case)

**Comment by soumyadeep2007** [comment]
Upstream, it is allowed on leaves. So let's keep it that way? Let's outlaw for only interior nodes?

File: src/backend/commands/tablecmds.c (hunk 2 above)
Comment (soumyadeep2007) [reply src/backend/commands/tablecmds.c]: Let's change the conditional to accommodate interior nodes in the tree:
```
bool is_interior_partition = rel_is_child_partition(rel_id) && !rel_is_leaf_partition(rel_id)
if(Gp_role != GP_ROLE_EXECUTE && (rel_is_partitioned(rel_id) || is_interior_partition))
....
```

File: src/test/regress/sql/create_index.sql [hunk 3]
```diff
@@ -805,6 +805,23 @@ DROP INDEX cwi_replaced_pkey;	-- Should fail; a constraint depends on it
...
+SUBPARTITION BY list(b)
+SUBPARTITION TEMPLATE (SUBPARTITION leaf VALUES (1))
+(PARTITION interior START(1) END(2) EVERY(1));
+-- disallowed
+CREATE UNIQUE INDEX ON root_part(a);
+ALTER TABLE root_part ADD PRIMARY KEY USING INDEX root_part_a_idx ;
+-- disallowed
+CREATE UNIQUE INDEX ON root_part_1_prt_interior(a);
+ALTER TABLE root_part_1_prt_interior ADD PRIMARY KEY USING INDEX root_part_1_prt_interior_a_idx1 ;
+-- allowed
```
Comment (berlin-ab) [inline src/test/regress/sql/create_index.sql]: why is it allowed in this case?

Comment (bhuvnesh2703) [reply src/test/regress/sql/create_index.sql]: because it's a leaf partition and this constraint need not be propagated any further down as opposed to the case of root / midlevel partition

**Comment by d** [comment]
> Upstream, it is allowed on leaves. So let's keep it that way? Let's outlaw for only interior nodes and root nodes?

TL;DR: it might not be worth the complexity, we might as well ban 'em all.

Long version: "leaf" partition is a misnomer, at least in Postgres. There is no multi-level partitioning, not in the Greenplum / Oracle sense. In Postgres, every table is either _partitioned_, or it's not. Perhaps it's better illustrated with an example:

```sql

CREATE TABLE foo (a int, b int, c int, d int) PARTITION BY RANGE(b);
CREATE TABLE primero PARTITION OF foo FOR VALUES FROM (0) TO (2);
CREATE TABLE segundo PARTITION OF foo FOR VALUES FROM (2) TO (4)
  PARTITION BY range(c);
CREATE TABLE alpha PARTITION OF segundo FOR VALUES FROM (10) TO (12);
CREATE TABLE beta PARTITION OF segundo FOR VALUES FROM (12) TO (14);
```

Here, there's no equi-depth tree. Tables `alpha`, `beta`, and `primero` are not partitioned. Table `foo` has two partitions: `primero` and `segundo`. Table `segundo` is itself partitioned into `alpha` and `beta`.

What' I am alluding to here is: what's cheap to check upstream (is relation `primero` partitioned?) might be too expensive to check in Greenplum, given the drastically divergent implementation of partitioning.

Jesse

File: src/backend/commands/tablecmds.c [hunk 4]
```diff
@@ -8852,6 +8852,16 @@ ATExecAddIndexConstraint(AlteredTableInfo *tab, Relation rel,
 	Assert(OidIsValid(index_oid));
 	Assert(stmt->isconstraint);
 
+	/*
+	 * Doing this on partitioned tables is not a simple feature to implement,
+	 * so let's punt for now.
+	 */
+	Oid rel_id = RelationGetRelid(rel);
+	if (Gp_role != GP_ROLE_EXECUTE && rel_is_parent(rel_id))
```
Comment (d) [inline src/backend/commands/tablecmds.c]: `rel_is_parent` won't cut it, as you're casting a wide net and catching inheritance tables as well

Comment (soumyadeep2007) [reply src/backend/commands/tablecmds.c]: Fixed. Thanks for the catch!

**Comment by berlin-ab** [comment]
This looks ready. Let's get it in and backported.

**Comment by berlin-ab** [comment]
What's the latest on this?

**Comment by soumyadeep2007** [comment]
w/ @bhuvnesh2703 
 
Took into account further feedback from @d on changing the location of the tests we added and the fact that we should have created the unique indexes in the tests with all columns of the partition key (in order to avoid the `alter add primary key using index` statement from prone to the failure: `ERROR:  PRIMARY KEY constraint must contain all columns in the partition key of relation`)

Waiting for the pipeline to go green, will merge, and then open a backport to 6.
